language/move-stdlib/sources/Capability.move

```diff
@@ -108,29 +133,39 @@ module Std::Capability {
             assert(exists<CapState<Feature>>(root_addr), Errors::invalid_state(EDELEGATE));
             assert(Vector::contains(&borrow_global<CapState<Feature>>(root_addr).delegates, &addr),
                    Errors::invalid_state(EDELEGATE));
-            Cap<Feature>{root: root_addr}
+            root_addr
         } else {
             assert(exists<CapState<Feature>>(addr), Errors::not_published(ECAP));
-            Cap<Feature>{root: addr}
+            addr
         }
     }
 
-    /// Registers a delegation relation.
-    public fun delegate<Feature>(cap: Cap<Feature>, to: &signer)
+    /// Returns the root address associated with the given capability token. Only the owner
+    /// of the feature can do this.
+    public fun root_addr<Feature>(cap: Cap<Feature>, _feature_witness: &Feature): address {
+        cap.root
+    }
+
+    /// Returns the root address associated with the given linear capability token.
+    public fun linear_root_addr<Feature>(cap: LinearCap<Feature>, _feature_witness: &Feature): address {
+        cap.root
+    }
+
+    /// Registers a delegation relation. If the relation already exists, this function does
+    /// nothing.
+    public fun delegate<Feature>(cap: Cap<Feature>, _feature_witness: &Feature, to: &signer)
     acquires CapState {
         let addr = Signer::address_of(to);
-        assert(!exists<CapDelegateState<Feature>>(addr), Errors::already_published(EDELEGATE));
-        assert(exists<CapState<Feature>>(cap.root), Errors::invalid_state(ECAP));
+        if (exists<CapDelegateState<Feature>>(addr)) return;
```

Comment:
I prefer failing in cases like these because failures are easy for clients to diagnose (they get a nice error code with documented mitigation steps), whereas silently doing nothing is much harder to identify.

Reply:
There are different tradeoffs here. For the revoke function, Tim and I discussed that it would be more useful to be able to call this without a concern of failure, so one can safely and without potentially aborting always ensure something is revoked. I thought delegate should then behave symmetric. We can certainly iterate on this once we have more use cases. I added a TODO.